modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/statechannel/StateChannelValidator.scala

```diff
@@ -34,20 +34,40 @@ trait StateChannelValidator[F[_]] {
   def validate(
     stateChannelOutput: StateChannelOutput,
     snapshotOrdinal: SnapshotOrdinal,
-    staked: Balance
+    snapshotFeesInfo: SnapshotFeesInfo
   )(implicit hasher: Hasher[F]): F[StateChannelValidationErrorOr[StateChannelOutput]]
+
   def validateHistorical(
     stateChannelOutput: StateChannelOutput,
     snapshotOrdinal: SnapshotOrdinal,
-    staked: Balance
+    snapshotFeesInfo: SnapshotFeesInfo
   )(
     implicit hasher: Hasher[F]
   ): F[StateChannelValidationErrorOr[StateChannelOutput]]
-
 }
 
 object StateChannelValidator {
 
+  def validateIfAddressAlreadyUsed(
+    metagraphId: Address,
+    existingFeesAddresses: Map[Address, Set[Address]],
+    maybeAddress: Option[Address]
+  ): StateChannelValidationErrorOr[Unit] = maybeAddress match {
+    case None => ().validNec
+    case Some(address) =>
+      val metagraphAddresses = existingFeesAddresses.getOrElse(metagraphId, List.empty)
+      if (metagraphAddresses.toList.contains(address)) {
+        ().validNec
+      } else {
+        val allAddresses = existingFeesAddresses.values.flatten.toList
```

Comment:
The problem with such an approach is that you always iterate over the whole `Map` to construct the flattened `List` and then iterate over that flattened `List` again. We will have a lot of addresses there in the future so I think you can just check the `Map[Address, Set[Address]]` directly without making it a flattened list. With such an approach you stop iterating over `Map` when you find the address so it is optimized:

```scala
val existsInAllAddresses = existingFeesAddresses.exists { case (_, addresses) => addresses.contains(address)}
          
if (existsInAllAddresses) {
  AddressAlreadyInUse.invalidNec
} else {
  stateChannelOutput.snapshotBinary.validNec
}
```

Reply:
Cool, thanks =]